File: core/plugin/__DEV/Python/n_gfUtilAngleScalarMath.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfUtilAngularScalarMath node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.

Requirements:
    Maya 2017 or above.

Todo:
    * Convert angle values to radians.

This code supports Pylint. Rc file in project.
"""

import math
import maya.api._OpenMaya_py2 as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=invalid-name, unnecessary-pass
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=invalid-name
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class AngularScalarMath(om2.MPxNode):
    """ Main class of gfUtilAngularScalarMath node. """

    kNodeName = ""
    kNodeClassify = ""
    kNodeID = ""

    inAngle = om2.MObject()
    inScalar = om2.MObject()
    inOperation = om2.MObject()
    outAngle = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return AngularScalarMath()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to AngularScalarMath class. Instances of AngularScalarMath will use these attributes to create plugs
        for use in the compute() method.
        """
        uAttr = om2.MFnUnitAttribute()
        nAttr = om2.MFnNumericAttribute()
        eAttr = om2.MFnEnumAttribute()

        AngularScalarMath.inAngle = uAttr.create("angle", "angle", om2.MFnUnitAttribute.kAngle, 0.0)
        INPUT_ATTR(uAttr)

        AngularScalarMath.inScalar = nAttr.create("scalar", "scalar", om2.MFnNumericData.kDouble, 0.0)
        INPUT_ATTR(nAttr)

        AngularScalarMath.inOperation = eAttr.create("operation", "operation", 0)
        eAttr.addField("No Operation", 0)
        eAttr.addField("Add", 1)
        eAttr.addField("Subtract", 2)
        eAttr.addField("Multiply", 3)
        eAttr.addField("Divide", 4)
        eAttr.addField("Power", 5)
        eAttr.addField("Min", 6)
        eAttr.addField("Max", 7)
        INPUT_ATTR(eAttr)

        AngularScalarMath.outAngle = uAttr.create("outAngle", "oa", om2.MFnUnitAttribute.kAngle, 0.0)
        OUTPUT_ATTR(uAttr)

        AngularScalarMath.addAttribute(AngularScalarMath.inOperation)
        AngularScalarMath.addAttribute(AngularScalarMath.inAngle)
        AngularScalarMath.addAttribute(AngularScalarMath.inScalar)
        AngularScalarMath.addAttribute(AngularScalarMath.outAngle)
        AngularScalarMath.attributeAffects(AngularScalarMath.inAngle, AngularScalarMath.outAngle)
        AngularScalarMath.attributeAffects(AngularScalarMath.inScalar, AngularScalarMath.outAngle)
        AngularScalarMath.attributeAffects(AngularScalarMath.inOperation, AngularScalarMath.outAngle)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=no-self-use
        if plug != AngularScalarMath.outAngle:
            return om2.kUnknownParameter

        angle = dataBlock.inputValue(AngularScalarMath.inAngle).asAngle().asDegrees()
        scalar = dataBlock.inputValue(AngularScalarMath.inScalar).asDouble()
        operation = dataBlock.inputValue(AngularScalarMath.inOperation).asShort()

        outAngleHandle = dataBlock.outputValue(AngularScalarMath.outAngle)

        if operation == 0:
            outAngleHandle.setMAngle(om2.MAngle(angle, om2.MAngle.kDegrees))
        elif operation == 1:
            outAngle = angle + scalar
            outAngleHandle.setMAngle(om2.MAngle(outAngle, om2.MAngle.kDegrees))
        elif operation == 2:
            outAngle = angle - scalar
            outAngleHandle.setMAngle(om2.MAngle(outAngle, om2.MAngle.kDegrees))
        elif operation == 3:
            outAngle = angle * scalar
            outAngleHandle.setMAngle(om2.MAngle(outAngle, om2.MAngle.kDegrees))
        elif operation == 4:
            if scalar != 0.0:
                outAngle = angle / scalar
            else:
                outAngle = 9999.999
            outAngleHandle.setMAngle(om2.MAngle(outAngle, om2.MAngle.kDegrees))
        elif operation == 5:
            outAngle = math.pow(angle, scalar)
            outAngleHandle.setMAngle(om2.MAngle(outAngle, om2.MAngle.kDegrees))
        elif operation == 6:
            outAngle = min(angle, scalar)
            outAngleHandle.setMAngle(om2.MAngle(outAngle, om2.MAngle.kDegrees))
        elif operation == 7:
            outAngle = max(angle, scalar)
            outAngleHandle.setMAngle(om2.MAngle(outAngle, om2.MAngle.kDegrees))

        outAngleHandle.setClean()

File: plugin/maya/nodes/source/n_gfUtilAngleScalarMath.cpp
#include "headers/n_gfUtilAngleScalarMath.h"

// Configure a input attribute.
#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

// Configure a output attribute.
#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\


// Constructor.
AngularScalarMath::AngularScalarMath() {}

// Destructor.
AngularScalarMath::~AngularScalarMath() {}

MObject AngularScalarMath::inAngle;
MObject AngularScalarMath::inScalar;
MObject AngularScalarMath::inOperation;
MObject AngularScalarMath::outAngle;

void* AngularScalarMath::creator(){
    // Maya creator function.
    return new AngularScalarMath();
}

MStatus AngularScalarMath::initialize(){
    /*
    Defines the set of attributes for this node. The attributes declared in this function are assigned
    as static members to AngularScalarMath class. Instances of AngularScalarMath will use these attributes to create plugs
    for use in the compute() method.
    */
    MStatus status;
    MFnUnitAttribute uAttr;
    MFnNumericAttribute nAttr;
    MFnEnumAttribute eAttr;

    inAngle = uAttr.create("angle", "angle", MFnUnitAttribute::kAngle, 0.0, &status);
    INPUT_ATTR(uAttr);

    inScalar = nAttr.create("scalar", "scalar", MFnNumericData::kDouble, 0.0, &status);
    INPUT_ATTR(nAttr);

    inOperation = eAttr.create("operation", "operation", 0, &status);
    eAttr.addField("No Operation", 0);
    eAttr.addField("Add", 1);
    eAttr.addField("Subtract", 2);
    eAttr.addField("Multiply", 3);
    eAttr.addField("Divide", 4);
    eAttr.addField("Power", 5);
    eAttr.addField("Min", 6);
    eAttr.addField("Max", 7);
    INPUT_ATTR(eAttr);

    outAngle = uAttr.create("outAngle", "oa", MFnUnitAttribute::kAngle, 0.0, &status);
    OUTPUT_ATTR(uAttr);

    addAttribute(inOperation);
    addAttribute(inAngle);
    addAttribute(inScalar);
    addAttribute(outAngle);
    attributeAffects(inOperation, outAngle);
    attributeAffects(inAngle, outAngle);
    attributeAffects(inScalar, outAngle);

    return status;
}

MStatus AngularScalarMath::compute(const MPlug& plug, MDataBlock& dataBlock){
    /*
    Node computation method:
        * plug is a connection point related to one of our node attributes (either an input or an output).
        * dataBlock contains the data on which we will base our computations.
    */
    if (plug != outAngle)
        return MStatus::kUnknownParameter;

    double angle = dataBlock.inputValue(inAngle).asAngle().asDegrees();
    double scalar = dataBlock.inputValue(inScalar).asDouble();
    short operation = dataBlock.inputValue(inOperation).asShort();

    MDataHandle outAngleHandle = dataBlock.outputValue(outAngle);
    double resAngle;

    switch (operation)
    {
    case 0:
        outAngleHandle.setMAngle(MAngle(angle, MAngle::kDegrees));
        break;
    case 1:
        resAngle = angle + scalar;
        outAngleHandle.setMAngle(MAngle(resAngle, MAngle::kDegrees));
        break;
    case 2:
        resAngle = angle - scalar;
        outAngleHandle.setMAngle(MAngle(resAngle, MAngle::kDegrees));
        break;
    case 3:
        resAngle = angle * scalar;
        outAngleHandle.setMAngle(MAngle(resAngle, MAngle::kDegrees));
        break;
    case 4:
        if (scalar != 0.0)
            resAngle = angle + scalar;
        else
            resAngle = 9999.999;
        outAngleHandle.setMAngle(MAngle(resAngle, MAngle::kDegrees));
        break;
    case 5:
        resAngle = pow(angle, scalar);
        outAngleHandle.setMAngle(MAngle(resAngle, MAngle::kDegrees));
        break;
    case 6:
        resAngle = min(angle, scalar);
        outAngleHandle.setMAngle(MAngle(resAngle, MAngle::kDegrees));
        break;
    case 7:
        resAngle = max(angle, scalar);
        outAngleHandle.setMAngle(MAngle(resAngle, MAngle::kDegrees));
        break;
    }
    outAngleHandle.setClean();

    return MStatus::kSuccess;
}